src/languages/janet/emit.ts

```diff
@@ -0,0 +1,170 @@
+  }
+
+  function joinNodes(delim: TokenTree, exprs: readonly IR.Node[]) {
+    return joinTrees(
+      delim,
+      exprs.map((x) => emit(x)),
+    );
+  }
+
+  /**
+   * Emits the expression.
+   * @param expr The expression to be emited.
+   * @returns  Token tree corresponding to the expression.
+   */
+  function emit(e: IR.Node): TokenTree {
+    switch (e.kind) {
+      case "Block":
+        return emitMultiNode(e);
+      case "While":
+        return ["(", "while", emit(e.condition), emitMultiNode(e.body), ")"];
+      case "ForEach":
+        return [
+          "(",
+          "each",
+          emit(e.variable),
+          emit(e.collection),
+          emitMultiNode(e.body),
+          ")",
+        ];
+      case "ForRange": {
+        const varName = e.variable === undefined ? "_" : emit(e.variable);
+        return isInt(1n)(e.increment)
+          ? [
+              "(",
+              "for",
+              varName,
+              emit(e.start),
+              emit(e.end),
+              emitMultiNode(e.body),
+              ")",
+            ]
+          : [
+              "(",
+              "loop",
+              "[",
+              varName,
+              ":range",
+              "[",
+              emit(e.start),
+              emit(e.end),
+              emit(e.increment),
+              "]",
+              "]",
+              emitMultiNode(e.body),
+              ")",
+            ];
+      }
+      case "If":
+        return [
+          "(",
+          "if",
+          emit(e.condition),
+          emitMultiNode(e.consequent, true),
+          e.alternate === undefined ? [] : emitMultiNode(e.alternate, true),
+          ")",
+        ];
+      case "VarDeclarationWithAssignment": {
+        const assignment = e.assignment;
+        if (assignment.kind !== "Assignment") {
+          throw new EmitError(
+            e,
+            `Declaration cannot contain ${assignment.kind}`,
+          );
+        }
+        const assignKeyword =
+          assignment.expr.kind === "Identifier" && assignment.expr.builtin
+            ? "def"
+            : "var";
+        return [
+          "(",
+          assignKeyword,
+          emit(assignment.variable),
+          emit(assignment.expr),
+          ")",
+        ];
+      }
+      case "Assignment":
+        return ["(", "set", emit(e.variable), emit(e.expr), ")"];
+      case "Identifier":
+        return e.name;
+      case "Text":
+        return emitJanetText(e.value);
+      case "Integer":
+        return emitIntLiteral(e, {
+          10: ["", ""],
+          16: ["0x", ""],
+          36: ["36r", ""],
+        });
+      case "FunctionCall":
+        return ["(", emit(e.func), e.args.map((x) => emit(x)), ")"];
+      case "MutatingInfix": {
+        const incOrDec =
+          (e.name === "+" || e.name === "-") && isInt(1n)(e.right);
+        if (incOrDec) {
+          return ["(", e.name === "+" ? "++" : "--", emit(e.variable), ")"];
+        }
```

Comment:
This is done using a plugin for js, maybe we want to be consistent? However, the plugin creates a postfix (should it be prefix?) node, which you are not emiting.

Reply:
Gotcha. I can add prefix/postfix in order to work with a plugin if need be. I left prefix and infix undefined just for simplicity. Janet is Lisp-like and doesn't have any precedence rules, so it seemed the simplest just to use FunctionCall for everything since its naturally variadic.

Reply:
At some point I'll probably go back and do the same with golfscript since theres no real reason for a distinction between prefix and infix in a polish notation language.

Reply:
Yes, adding I agree that adding the prefix/infix/postfix nodes to Janet doesn't make much sense. But I think the logic of detecting that something is in fact inc/dec op should be centralised.